ntex-bytes/src/buf/pool.rs

```diff
@@ -0,0 +1,269 @@
+    fn take(&self, len: usize) -> Self::TakeFuture;
+}
+
+impl<T> BufferSource for &'_ T where T: BufferSource {
+    type Owned = T::Owned;
+    type Accumulator = T::Accumulator;
+    type TakeFuture = T::TakeFuture;
+
+    fn take(&self, len: usize) -> Self::TakeFuture {
+        (**self).take(len)
+    }
+}
+
+/// A sink of buffers. Dropped buffers will automatically return to this sink.
+pub trait BufferSink {
+    /// The type of buffer returned by this buffer source that owns its contents.
+    type Owned: Owned;
+
+    /// The type of backing buffer that is returned to this pool when the last owned or shared buffer
+    /// holding on to said backing buffer is dropped.
+    type Backing;
+
+    /// Returns the given backing buffer to the pool. This is intended to be invoked from
+    /// the `Drop` impls of this pool's owned and shared buffer types.
+    fn put_back(&self, backing: Self::Backing);
+}
+
+/// Convenience trait for a type that is a buffer pool, ie a type that is both a [`BufferSource`]
+/// and a [`BufferSink`].
+pub trait BufferPool: BufferSource + BufferSink<Owned = <Self as BufferSource>::Owned> {}
+
+impl<T> BufferPool for T where T: BufferSource + BufferSink<Owned = <Self as BufferSource>::Owned> {}
+
+impl<T> BufferSink for Rc<T>
+where
+    T: ?Sized + BufferSink,
+{
+    type Backing = T::Backing;
+    type Owned = T::Owned;
+
+    fn put_back(&self, backing: Self::Backing) {
+        (&**self).put_back(backing);
+    }
+}
+
+impl<T> BufferSink for Arc<T>
+where
+    T: BufferSink,
+{
+    type Backing = T::Backing;
+    type Owned = T::Owned;
+
+    fn put_back(&self, backing: Self::Backing) {
+        (&**self).put_back(backing);
+    }
+}
+
+/// A [`BufferSource`] that always yields new `BytesMut` without blocking.
+#[derive(Clone, Copy, Debug)]
+pub struct BytesPool;
+
+impl BufferSource for BytesPool {
+    type Owned = BytesMut;
+    type Accumulator = BytesAccumulator;
+    type TakeFuture = future::Ready<Self::Owned>;
+
+    fn take(&self, len: usize) -> Self::TakeFuture {
+        future::ready(BytesMut::with_capacity(len))
+    }
+}
+
+impl BufferSink for BytesPool {
+    type Backing = ();
+    type Owned = <Self as BufferSource>::Owned;
+
+    fn put_back(&self, _backing: Self::Backing) {
+    }
+}
+
+/// A buffer that owns a particular range of its backing buffer.
+///
+/// An `Owned` tracks what part of itself has been filled with data.
+/// The filled region can be accessed with [`Owned::filled`].
+/// The unfilled region can be accessed with [`Owned::unfilled_mut`], and bytes can be moved from the start of this region
+/// into the end of the filled region with [`Owned::fill`].
+///
+/// An `Owned` can be subdivided into smaller `Owned`s with `split_to` that each own
+/// smaller splits of the backing buffer.
+///
+/// An `Owned` is not `Clone`. It can be converted to a [`Shared`] which is, via [`Owned::freeze`]
+pub trait Owned {
```

Comment:
do we need `len()`, `capacity()`?

Reply:
I don't think so.

Reply:
why? so generically it would be not possible to know len and cap?

Reply:
Yes. What do you think the caller would need that for that isn't already served by knowing the filled and unfilled lengths individually?

Reply:
i dont really understand all this filled and unfilled concepts. to me it is the same as what does vec or bytesmut. i want public api be as close to vec and bytes as possible. i am not really interested in inventing new names for same concepts.

Reply:
They're not names I invented. Look at https://docs.rs/tokio/1.13.0/tokio/io/struct.ReadBuf.html